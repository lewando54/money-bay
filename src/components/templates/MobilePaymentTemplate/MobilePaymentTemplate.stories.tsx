/* eslint-disable @typescript-eslint/consistent-type-assertions */
import { StoryObj, Meta } from '@storybook/react-native'

import MobilePaymentTemplate from './MobilePaymentTemplate'

const meta = {
	component: MobilePaymentTemplate,
	argTypes: {
        // To write
	}
} satisfies Meta<typeof MobilePaymentTemplate>

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
	args: {
        // To write
	}
}
