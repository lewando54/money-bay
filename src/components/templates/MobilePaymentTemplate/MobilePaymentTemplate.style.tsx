import { Dimensions, StyleSheet } from 'react-native'

import globalThemeStyle, { MAIN_DARK } from '@styling/GlobalTheme.style'
const { width, height } = Dimensions.get('screen')
export default StyleSheet.create({
    mainContainer: {
        padding: width * 0.0533,
        gap: height * 0.0184
    },
    balanceText: {
        ...globalThemeStyle.text_Regular,
        color: MAIN_DARK,
        fontSize: 12
    },
    amountInputWrapper: {
        flexDirection: 'row',
        gap: 10,
        alignItems: 'center',
        justifyContent: 'flex-start',
        flexGrow: 0
    },
    moneyInputContainer: {
        flexDirection: 'row',
        justifyContent: 'flex-start',
        alignItems: 'flex-end',
        width: 'auto'
    },
    bigMoneyText: {
        ...globalThemeStyle.text_Medium,
        color: MAIN_DARK,
        fontSize: 28
    },
    smallMoneyText: {
        ...globalThemeStyle.text_Medium,
        color: MAIN_DARK,
        fontSize: 16
    }
})
