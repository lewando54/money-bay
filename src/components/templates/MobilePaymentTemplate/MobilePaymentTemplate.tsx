import React, { useRef } from 'react'
import { Platform, TextInput, View } from 'react-native'

import { CustomText } from '@atoms'
import { Button, UniversalContainer } from '@atoms'
import { PhoneInput } from '@molecules'
import { TCurrency } from '@utils/currency'

import mobilePaymentTemplateStyle from './MobilePaymentTemplate.style'

interface IMobilePaymentTemplateProps {
    balance?: number
    currency?: TCurrency
    phoneValue?: string
    wholeValue?: string
    decimalValue?: string
    onPhoneChange?: (newValue: string) => void
    onWholeChange?: (newValue: string) => void
    onDecimalChange?: (newValue: string) => void
    onSubmit?: () => void
    testID?: string
}

export default function MobilePaymentTemplate({
    balance=4863.27,
    currency={iso: 'usd', symbol: '$', left: true},
    phoneValue,
    wholeValue,
    decimalValue,
    onPhoneChange,
    onWholeChange,
    onDecimalChange,
    onSubmit,
    testID
}: IMobilePaymentTemplateProps){

    const amountWholeRef = useRef<TextInput>()
    const amountDecimalRef = useRef<TextInput>()

    const onAmountWholeChangeHandle = (newValue: string) => {
        if(newValue.includes(',') || newValue.includes('.')){
            newValue = newValue.replace(',', '')
            newValue = newValue.replace('.', '')
            amountWholeRef.current.setNativeProps({
                text: newValue
            })
            amountDecimalRef.current.focus()
        }
        else if(newValue.includes('-')){
            newValue = newValue.replace('-', '')
            amountWholeRef.current.setNativeProps({
                text: newValue
            })
        }
        if(onWholeChange){
            onWholeChange(newValue)
        }
    }

    const onAmountDecimalChangeHandle = (newValue: string) => {
        if(newValue.includes(',') || newValue.includes('.') || newValue.includes('-')){
            newValue = newValue.replace(',', '')
            newValue = newValue.replace('.', '')
            newValue = newValue.replace('-', '')
            amountDecimalRef.current.setNativeProps({
                text: newValue
            })
        }
        if(onDecimalChange){
            onDecimalChange(newValue)
        }
    }

    return (
        <View style={mobilePaymentTemplateStyle.mainContainer} testID={testID}>
            <PhoneInput
                value={phoneValue}
                onChange={onPhoneChange}
            />
            <CustomText style={mobilePaymentTemplateStyle.balanceText}>Your balance: {balance+' '+currency.iso.toLocaleUpperCase()}</CustomText>
            <View style={mobilePaymentTemplateStyle.amountInputWrapper}>
                <UniversalContainer style={mobilePaymentTemplateStyle.moneyInputContainer} castShadow={true}>
                    <CustomText style={[mobilePaymentTemplateStyle.bigMoneyText, {marginRight: 5}]}>$</CustomText>
                    <TextInput
                        value={wholeValue} 
                        ref={amountWholeRef}
                        placeholder={'0'}
                        onChangeText={onAmountWholeChangeHandle}
                        style={[mobilePaymentTemplateStyle.bigMoneyText, {flexGrow: 0}]}
                        keyboardType={Platform.select({android: 'numeric', ios: 'number-pad'})}
                        testID='wholeNumber'
                    />
                    <CustomText style={mobilePaymentTemplateStyle.bigMoneyText}>.</CustomText>
                    <TextInput
                        value={decimalValue}
                        maxLength={2}
                        placeholder={'00'}
                        onChangeText={onAmountDecimalChangeHandle}
                        ref={amountDecimalRef}
                        style={mobilePaymentTemplateStyle.smallMoneyText}
                        keyboardType={Platform.select({android: 'numeric', ios: 'number-pad'})}
                        testID='decimalNumber'
                    />
                </UniversalContainer>
                <CustomText style={mobilePaymentTemplateStyle.balanceText}>No fees</CustomText>
            </View>
            <Button onClick={onSubmit}>Confirm</Button>
        </View>
    )
}
