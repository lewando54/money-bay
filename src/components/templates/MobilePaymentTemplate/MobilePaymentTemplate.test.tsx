import React from 'react'
import renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react-native'
import MobilePaymentTemplate from './MobilePaymentTemplate'

describe('MobilePaymentTemplate', () => {
    it('should render correctly', () => {
        const tree = renderer.create(<MobilePaymentTemplate />)
        expect(tree).toMatchSnapshot()
    })

    it('should remove [".", ",", "-"] correctly', () => {
        const onChange = jest.fn()
        const { getByTestId } = render(<MobilePaymentTemplate onDecimalChange={onChange} onWholeChange={onChange}/>)
        fireEvent.changeText(getByTestId('wholeNumber'), 'test,')
        expect(onChange).toBeCalledWith('test')
        fireEvent.changeText(getByTestId('wholeNumber'), 'test.')
        expect(onChange).toBeCalledWith('test')
        fireEvent.changeText(getByTestId('wholeNumber'), 'test-')
        expect(onChange).toBeCalledWith('test')
        fireEvent.changeText(getByTestId('decimalNumber'), 'test,')
        expect(onChange).toBeCalledWith('test')
        fireEvent.changeText(getByTestId('decimalNumber'), 'test.')
        expect(onChange).toBeCalledWith('test')
        fireEvent.changeText(getByTestId('decimalNumber'), 'test-')
        expect(onChange).toBeCalledWith('test')
    })
})
