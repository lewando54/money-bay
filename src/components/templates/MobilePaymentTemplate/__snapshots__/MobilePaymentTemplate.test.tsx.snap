// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MobilePaymentTemplate should render correctly 1`] = `
<View
  style={
    {
      "gap": 24.5456,
      "padding": 39.975,
    }
  }
>
  <View
    style={
      [
        {
          "backgroundColor": "white",
          "borderRadius": 10,
          "height": "auto",
          "padding": 10,
          "width": "100%",
        },
        {
          "elevation": 20,
          "shadowColor": "#16196670",
          "shadowOffset": {
            "height": 6,
            "width": 0,
          },
          "shadowOpacity": 0.08,
          "shadowRadius": 16,
        },
        [
          {
            "paddingHorizontal": 20,
          },
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 7,
          },
        ],
      ]
    }
    testID="universal-container"
  >
    <View
      style={
        {
          "alignItems": "center",
          "flexDirection": "row",
          "gap": 7,
        }
      }
    >
      <Image
        source={1}
        style={
          {
            "resizeMode": "contain",
            "width": 20,
          }
        }
      />
      <Text
        style={
          [
            {
              "color": "#1B1D4D",
              "fontFamily": "MulishRegular",
              "fontSize": 14,
            },
            {
              "color": "#4C4C60",
              "fontFamily": "MulishRegular",
              "fontSize": 14,
            },
          ]
        }
      >
        +17
      </Text>
    </View>
    <Text
      style={
        [
          {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 14,
          },
          {
            "color": "#4C4C60",
            "fontFamily": "MulishRegular",
            "fontSize": 14,
          },
        ]
      }
    >
      |
    </Text>
    <TextInput
      keyboardType="number-pad"
      maxLength={9}
      onChangeText={[Function]}
      placeholder="xxxxxxxxx"
      style={
        [
          {
            "width": 100,
          },
          {
            "color": "#4C4C60",
            "fontFamily": "MulishRegular",
            "fontSize": 14,
          },
        ]
      }
    />
  </View>
  <Text
    style={
      [
        {
          "color": "#1B1D4D",
          "fontFamily": "MulishRegular",
          "fontSize": 14,
        },
        {
          "color": "#1B1D4D",
          "fontFamily": "MulishRegular",
          "fontSize": 12,
        },
      ]
    }
  >
    Your balance: 
    4863.27 USD
  </Text>
  <View
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "flexGrow": 0,
        "gap": 10,
        "justifyContent": "flex-start",
      }
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "white",
            "borderRadius": 10,
            "height": "auto",
            "padding": 10,
            "width": "100%",
          },
          {
            "elevation": 20,
            "shadowColor": "#16196670",
            "shadowOffset": {
              "height": 6,
              "width": 0,
            },
            "shadowOpacity": 0.08,
            "shadowRadius": 16,
          },
          {
            "alignItems": "flex-end",
            "flexDirection": "row",
            "justifyContent": "flex-start",
            "width": "auto",
          },
        ]
      }
      testID="universal-container"
    >
      <Text
        style={
          [
            {
              "color": "#1B1D4D",
              "fontFamily": "MulishRegular",
              "fontSize": 14,
            },
            [
              {
                "color": "#1B1D4D",
                "fontFamily": "MulishMedium",
                "fontSize": 28,
              },
              {
                "marginRight": 5,
              },
            ],
          ]
        }
      >
        $
      </Text>
      <TextInput
        keyboardType="number-pad"
        onChangeText={[Function]}
        placeholder="0"
        style={
          [
            {
              "color": "#1B1D4D",
              "fontFamily": "MulishMedium",
              "fontSize": 28,
            },
            {
              "flexGrow": 0,
            },
          ]
        }
        testID="wholeNumber"
      />
      <Text
        style={
          [
            {
              "color": "#1B1D4D",
              "fontFamily": "MulishRegular",
              "fontSize": 14,
            },
            {
              "color": "#1B1D4D",
              "fontFamily": "MulishMedium",
              "fontSize": 28,
            },
          ]
        }
      >
        .
      </Text>
      <TextInput
        keyboardType="number-pad"
        maxLength={2}
        onChangeText={[Function]}
        placeholder="00"
        style={
          {
            "color": "#1B1D4D",
            "fontFamily": "MulishMedium",
            "fontSize": 16,
          }
        }
        testID="decimalNumber"
      />
    </View>
    <Text
      style={
        [
          {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 14,
          },
          {
            "color": "#1B1D4D",
            "fontFamily": "MulishRegular",
            "fontSize": 12,
          },
        ]
      }
    >
      No fees
    </Text>
  </View>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={false}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
        "width": "100%",
      }
    }
  >
    <ViewManagerAdapter_ExpoLinearGradient
      proxiedProperties={
        {
          "colors": [
            4288142079,
            4279654301,
          ],
          "endPoint": [
            1,
            0.5,
          ],
          "locations": undefined,
          "startPoint": [
            0,
            0.5,
          ],
        }
      }
      style={
        [
          {
            "borderRadius": 10,
            "padding": 15,
          },
          false,
          false,
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#1B1D4D",
              "fontFamily": "MulishRegular",
              "fontSize": 14,
            },
            [
              {
                "fontFamily": "MulishSemiBold",
                "fontSize": 16,
                "textAlign": "center",
              },
              {
                "color": "white",
              },
              false,
              false,
            ],
          ]
        }
      >
        Confirm
      </Text>
    </ViewManagerAdapter_ExpoLinearGradient>
  </View>
</View>
`;
